pkg/cluster/mutex.go

```diff
@@ -31,20 +32,30 @@ type Mutex interface {
 }
 
 type mutex struct {
+	// concurrency.Mutex is a session level mutex, so sync.Mutex is
+	// required to make it goroutine safe
+	lock    sync.Mutex
 	m       *concurrency.Mutex
 	timeout time.Duration
 }
 
 func (m *mutex) Lock() error {
+	m.lock.Lock()
+
 	ctx, cancel := context.WithTimeout(context.Background(), m.timeout)
 	defer cancel()
 
-	return m.m.Lock(ctx)
+	err := m.m.Lock(ctx)
+	if err != nil {
+		m.lock.Unlock()
```

Comment:
If we make it safest, we should use `defer` statement to unlock if the panic happened perhaps from `m.m.Lock`. Otherwise, it will cause the local lock won't be released.

Reply:
when `m.m.Lock` succeeds, we should not unlock `m.lock`, it will be unlocked in `m.Unlock`.

Reply:
based on the implementation of etcd mutex, we need to linearize the execution of Lock and Unlock by ourself...

Reply:
> when m.m.Lock succeeds, we should not unlock m.lock, it will be unlocked in m.Unlock.

Sure, we can know it to panic or not in recovery function. The point here is `m.m.Lock` has a change to panic, then the local mutex will be locked forever. we could use a defer function to avoid it.

Reply:
Thanks for the comment, updated.